nixos.system.boot.kernel: stronger constraints for kernelParam type

cab404 opened the pull request:
Checks whether all spaces are inside double quotes, thus ensuring that one
string parses as no more than one kernel param.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Users often add spaces in expressions like "a = 12", and having them parsed as 3 separate params is not user friendly.

###### Things done

kernelParams type changed to a string match.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

pennae commented on nixos/modules/system/boot/kernel.nix:
```diff
@@ -83,7 +83,10 @@ in
 
     boot.kernelParams = mkOption {
-      type = types.listOf types.str;
+      type = types.listOf (types.strMatching ''([^" ]*|"[^"]*")*'' // {
```
this should check for all code points `isspace()` detects as spaces. maybe for all spaces in general, better to be a little overzealous in having people add quotes than create bad command lines

cab404 replied:
sounds reasonable, will change to \s when back home

cab404 commented:
It actually contains some breaking changes (e.g some of the old configs won't build), so I will need to update release notes.

cab404 committed:
nixos.system.boot.kernel: stronger constraints for kernelParam type

Checks whether all spaces are inside double quotes, thus ensuring that one
string parses as no more than one kernel param.

Co-authored-by: pennae <82953136+pennae@users.noreply.github.com>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/575

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/403

Artturin commented on nixos/doc/manual/release-notes/rl-2111.section.md:
```diff
@@ -171,7 +171,6 @@ pt-services.clipcat.enable).
 
 - `programs.neovim.runtime` switched to a `linkFarm` internally, making it impossible to use wildcards in the `source` argument.
 
-<<<<<<< HEAD
```
You forgot this

nrdxp reviewed (approved):
Seems to work as intended when evaluating a system from this branch of nixpkgs. Just rebase and you've got my approval to merge :+1:

nrdxp commented:
There is no demerge button unfortunately, but I'll gladly merge a doc PR :sweat_smile: 
Sorry, I guess I got a little antsy :sweat:

cab404 commented:
> There is no demerge button unfortunately, but I'll gladly merge a doc PR sweat_smile Sorry, I guess I got a little antsy sweat

#141818 :heart:
